pkg/scheduler/cache/cache.go

```diff
@@ -176,6 +184,251 @@ type imageState struct {
 	nodes sets.Set[string]
 }
 
+// MemberMatcher represents a matcher for a child node (exact for regex)
+type MemberMatcher struct {
+	Exact string // Exact string match
+	Regex string // Regex match
+}
+
+// Matches checks if the node name matches the MemberMatcher
+func (mm *MemberMatcher) Matches(name string) bool {
+	if mm.Exact != "" {
+		result := mm.Exact == name
+		klog.V(4).Infof("Matching node %s with exact %s: %v", name, mm.Exact, result)
+		return result
+	}
+	if mm.Regex != "" {
+		r, err := regexp.Compile(mm.Regex)
+		if err != nil {
+			klog.Errorf("Failed to compile regex %s: %v", mm.Regex, err)
+			return false
+		}
+		result := r.MatchString(name)
+		klog.V(4).Infof("Matching node %s with regex %s: %v", name, mm.Regex, result)
+		return result
+	}
+	return false
+}
+
+// NewExactMatcher creates a new MemberMatcher for exact matching
+func NewExactMatcher(exact string) *MemberMatcher {
+	return &MemberMatcher{Exact: exact}
+}
+
+// NewRegexMatcher creates a new MemberMatcher for regex matching
+func NewRegexMatcher(regex string) (*MemberMatcher, error) {
+	_, err := regexp.Compile(regex)
+	if err != nil {
+		klog.Errorf("Failed to compile regex %s: %v", regex, err)
+		return nil, err
+	}
+	return &MemberMatcher{Regex: regex}, nil
+}
+
+// HyperNodeTreeNode reprensent a single hypernode in the tree
+type HyperNodeTreeNode struct {
+	Name           string
+	Tier           int
+	Parent         *HyperNodeTreeNode
+	Members        sets.Set[*HyperNodeTreeNode]
+	MemberMatchers sets.Set[*MemberMatcher]
+	Nodes          sets.Set[*schedulingapi.NodeInfo]
+	NodeMatchers   sets.Set[*MemberMatcher]
+	HyperNodeRef   *networktopov1alpha1.HyperNode
+}
+
+func NewHyperNodeTreeNode(name string) *HyperNodeTreeNode {
+	return &HyperNodeTreeNode{
+		Name:           name,
+		Tier:           -1, // Initialize to an invalid value
+		Members:        sets.New[*HyperNodeTreeNode](),
+		MemberMatchers: sets.New[*MemberMatcher](),
+		Nodes:          sets.New[*schedulingapi.NodeInfo](),
+		NodeMatchers:   sets.New[*MemberMatcher](),
+	}
+}
+
+type HyperNodeSet sets.Set[*HyperNodeTreeNode]
+
+// HyperNodeForestBuilder helps in building the forest of hypernodes
+type HyperNodeForestBuilder struct {
+	nodes map[string]*HyperNodeTreeNode // Map to store nodes temporarily
+	roots sets.Set[string]              // Set of root node names
+}
+
+// NewHyperNodeForestBuilder initializes and returns a ForestBuilder
+func NewHyperNodeForestBuilder() *HyperNodeForestBuilder {
+	return &HyperNodeForestBuilder{
+		nodes: make(map[string]*HyperNodeTreeNode),
+		roots: sets.New[string](),
+	}
+}
+
+// insertHyperNodeMember inserts a member into a hypernode and updates the forest
+func insertHyperNodeMember(hntn *HyperNodeTreeNode, hnfb *HyperNodeForestBuilder, memberName string, memberNode *HyperNodeTreeNode, sc *SchedulerCache) {
```

Comment:
There is no distinction between the nodes in the leaf nodes and the hyperNodes of non-leaf nodes. What we expect is that the key stored in `HyperNodes` is the hyperNode, and the value is a list of all leaf nodes under the hyperNode.